File: src/net/merayen/elastic/ui/objects/components/DropDown.java
package net.merayen.elastic.ui.objects.components;

import net.merayen.elastic.ui.Draw;
import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.event.UIEvent;
import net.merayen.elastic.ui.event.MouseEvent.Button;
import net.merayen.elastic.ui.objects.top.menu.MenuList;
import net.merayen.elastic.ui.util.MouseHandler;
import net.merayen.elastic.util.Point;

public class DropDown extends UIObject {
	public float width = 100;
	private final float height = 20;
	private MouseHandler mouse_handler;
	private boolean open;
	public MenuList menu_list = new MenuList();

	@Override
	public void onInit() {
		menu_list.setHandler(new MenuList.Handler() {
			@Override
			public void onOutsideClick() {
				if(menu_list.getParent() != null)
					toggle();
			}
		});
		mouse_handler = new MouseHandler(this, Button.LEFT);
		mouse_handler.setHandler(new MouseHandler.Handler() {
			@Override
			public void onMouseDown(Point position) {
				toggle();
			}

			@Override
			public void onMouseOutsideDown(Point global_position) {
				//if(open)
				//	toggle();
			}
		});
	}

	@Override
	public void onDraw(Draw draw) {
		draw.setColor(0, 0, 0);
		draw.setStroke(1);
		draw.rect(0, 0, width, height);

		draw.rect(width - 20, 0, 20, height);
	}

	@Override
	public void onUpdate() {
		menu_list.getTranslation().y = height;
	}

	@Override
	public void onEvent(UIEvent e) {
		mouse_handler.handle(e);
	}

	private void toggle() {
		if(menu_list.getParent() == null)
			add(menu_list);
		else
			remove(menu_list);
	}
}

File: src/main/kotlin/net/merayen/elastic/ui/objects/components/DropDown.kt
package net.merayen.elastic.ui.objects.components

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.FlexibleDimension
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.event.MouseEvent
import net.merayen.elastic.ui.event.UIEvent
import net.merayen.elastic.ui.objects.contextmenu.ContextMenu
import net.merayen.elastic.ui.objects.contextmenu.ContextMenuItem
import net.merayen.elastic.util.Point

class DropDown(private val handler: Handler) : UIObject(), FlexibleDimension {
	interface Handler {
		fun onChange(selected: Item)
	}

	open class Item(val dropdownItem: UIObject, val contextMenuItem: ContextMenuItem)

	override var layoutWidth = 100f
	override var layoutHeight = 15f

	private val items = ArrayList<Item>()
	private var currentItem: Item? = null

	private val contextMenu = ContextMenu(this, MouseEvent.Button.LEFT)

	override fun onInit() {
		contextMenu.handler = object : ContextMenu.Handler {
			override fun onMouseDown(position: Point) {}

			override fun onSelect(item: ContextMenuItem?, position: Point) {
				var selected: Item? = null

				for (m in items)
					if (m.contextMenuItem === item)
						selected = m

				if (selected != null) {
					setViewItem(selected)
					handler.onChange(selected)
				}
			}
		}
	}

	override fun onDraw(draw: Draw) {
		draw.setColor(0, 0, 0)
		draw.setStroke(1f)
		draw.rect(0f, 0f, layoutWidth, layoutHeight)
	}

	fun addMenuItem(item: Item) {
		items.add(item)
		contextMenu.addMenuItem(item.contextMenuItem)
	}

	override fun onEvent(event: UIEvent) {
		contextMenu.handle(event)
	}

	fun setViewItem(item: Item) {
		val cItem = currentItem
		if(cItem != null)
			remove(cItem.dropdownItem)

		item.dropdownItem.translation.x = 5f
		item.dropdownItem.translation.y = 1f
		add(item.dropdownItem)

		currentItem = item
	}

	fun getItems() = items.toTypedArray()
}
